### PR TITLE
WIP Disk Transform / ApproxFunBase support

### PR DESCRIPTION
**Pull request by dlfivefifty**

### CODE REVIEW

**Comment by dlfivefifty** [comment]
@MikaelSlevinsky I believe there is a DiskHarmonic transform bug shown at the end of `test_disk.jl`:
```julia
m,ℓ = (2,4);
p1 = (r,θ) -> sqrt(2ℓ+2)*r^m*jacobip((ℓ-m)÷2, m, 0, 2r^2-1)*cos(m*θ)/sqrt(2π)
f = Fun((x,y) -> p1(sqrt(x^2+y^2), atan(y,x)), ChebyshevDisk());
@test f(0.1*cos(0.2),0.1*sin(0.2)) ≈ p1(0.1,0.2)
c = totensor(f.space, chop(f.coefficients,1E-12))
@test chebydiskeval(c, 0.1, 0.2) ≈ p1(0.1,0.2)


n = size(c,1); c = sqrt(2π)*pad(c,n,4n-3)
P = CDisk2CxfPlan(n)
d = P \ c
@test d[(ℓ-m)÷2, 2m+1] ≈ 0 # fails
@test d[(ℓ-m)÷2+1, 2m+1] ≈ 1
```

In particular, we have:
```julia
julia> c
7×25 Array{Float64,2}:
 -1.86358e-16   5.26847e-16  2.80209e-16  3.99688e-16  3.16228  5.26847e-16  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 -2.73611e-16   2.31244e-16  1.47033e-16  5.76776e-16  4.74342  2.31244e-16     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 -9.4797e-17   -3.4707e-17   2.58672e-17  9.27862e-17  1.58114  0.0             0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  1.3547e-18   -4.94907e-17  0.0          0.0          0.0      0.0             0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  0.0           0.0          0.0          0.0          0.0      0.0             0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  0.0           0.0          0.0          0.0          0.0      0.0          …  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  0.0           0.0          0.0          0.0          0.0      0.0             0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0

julia> d
7×25 Array{Float64,2}:
 -1.09431e-16   2.32317e-16  1.11288e-16   3.43552e-16   2.58199      …  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 -1.12033e-16   1.21614e-16  6.68733e-17   8.08986e-17   1.0             0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 -3.99699e-17  -1.27649e-17  1.19476e-17  -1.31428e-17  -9.74578e-17     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  5.79295e-19  -2.26243e-17  0.0           8.94155e-18   6.6304e-17      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  0.0           0.0          0.0          -6.59017e-18  -4.88679e-17     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  0.0           0.0          0.0           5.11734e-18   3.79464e-17  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  0.0           0.0          0.0          -8.51577e-18  -6.31467e-17     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
```

Can you double check this is indeed on your side?

**Comment by MikaelSlevinsky** [comment]
I'm not on that branch, but it seems like the `jacobip` parameters should be `(alpha,beta) = (0,m)` and the orthonormal cosines should be divided by only `sqrt(pi)`?

**Comment by MikaelSlevinsky** [comment]
Also, looking at the c_transforms.jl file, it seems to me like `*` and `\` with `CDisk2CxfPlan` should have the same sizes.

**Comment by MikaelSlevinsky** [comment]
I replaced the test function `f(x,y)` in [holomorphic.c](https://github.com/MikaelSlevinsky/FastTransforms/blob/master/examples/holomorphic.c) by
```c
double f(double r, double t) {return sqrt(10.0)*r*r*(4.0*r*r-3.0)*cos(2*t)/sqrt(M_PI);};
```
and changed the sampling [line 60](https://github.com/MikaelSlevinsky/FastTransforms/blob/master/examples/holomorphic.c#L60) to reflect polar coordinates `F[n+N*m] = f(r[n], theta[m]);` and it returns a `1.0` in the right place.

**Comment by dlfivefifty** [comment]
Got it! (I thought `jacobip` had the parameters reversed, guess not.):
```julia
m,ℓ = (2,4);
p1 = (r,θ) -> sqrt(2ℓ+2)*r^m*jacobip((ℓ-m)÷2, 0, m, 2r^2-1)*cos(m*θ)/sqrt(π)
f = Fun((x,y) -> p1(sqrt(x^2+y^2), atan(y,x)), ChebyshevDisk());
@test f(0.1*cos(0.2),0.1*sin(0.2)) ≈ p1(0.1,0.2)
c = totensor(f.space, chop(f.coefficients,1E-12))
@test chebydiskeval(c, 0.1, 0.2) ≈ p1(0.1,0.2)


n = size(c,1); c = sqrt(π)*pad(c,n,2n+1)
P = CDisk2CxfPlan(n)
d = P \ c
@test d[(ℓ-m)÷2, 2m+1] ≈ 0 atol = 1E-13
@test d[(ℓ-m)÷2+1, 2m+1] ≈ 1
```

**Comment by dlfivefifty** [comment]
Hmm I get SEGFAULTs once in a while:
```julia
signal (11): Segmentation fault: 11
in expression starting at no file:0
fftw_problem_destroy at /Users/sheehanolver/.julia/packages/FFTW/p7sLQ/deps/usr/lib/libfftw3.3.dylib (unknown line)
fftw_destroy_plan at /Users/sheehanolver/.julia/packages/FFTW/p7sLQ/deps/usr/lib/libfftw3.3.dylib (unknown line)
destroy_plan at /Users/sheehanolver/.julia/packages/FFTW/p7sLQ/src/fft.jl:255
jl_apply at /Users/sheehanolver/Projects/julia-1.1/src/./julia.h:1571 [inlined]
run_finalizer at /Users/sheehanolver/Projects/julia-1.1/src/gc.c:232
jl_gc_run_finalizers_in_list at /Users/sheehanolver/Projects/julia-1.1/src/gc.c:318
run_finalizers at /Users/sheehanolver/Projects/julia-1.1/src/gc.c:346 [inlined]
jl_gc_collect at /Users/sheehanolver/Projects/julia-1.1/src/gc.c:2892
jl_gc_pool_alloc at /Users/sheehanolver/Projects/julia-1.1/src/gc.c:1102
getindex at ./tuple.jl:24 [inlined]
iterate at ./tuple.jl:43 [inlined]
mapfoldl_impl at ./reduce.jl:43
mapfoldl_impl at ./reduce.jl:61 [inlined]
#mapfoldl#187 at ./reduce.jl:72 [inlined]
mapfoldl at ./reduce.jl:72 [inlined]
#mapreduce#191 at ./reduce.jl:205 [inlined]
mapreduce at ./reduce.jl:205 [inlined]
length at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/Multivariate/TensorSpace.jl:32 [inlined]
start at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/Multivariate/TensorSpace.jl:35 [inlined]
iterate at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/Multivariate/TensorSpace.jl:64 [inlined]
fromtensor at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/Multivariate/TensorSpace.jl:493
fromtensor at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/Multivariate/TensorSpace.jl:485 [inlined]
torectcfs at /Users/sheehanolver/Projects/MultivariateOrthogonalPolynomials.jl/src/Disk/Disk.jl:82
evaluate at /Users/sheehanolver/Projects/MultivariateOrthogonalPolynomials.jl/src/Disk/Disk.jl:88
evaluate at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/Fun.jl:216 [inlined]
Fun at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/Fun.jl:220
unknown function (ip: 0x12b1a4178)
#26 at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/constructors.jl:133
_all at ./reduce.jl:663
unknown function (ip: 0x12b1a3c62)
#all#561 at ./reducedim.jl:661
unknown function (ip: 0x12b1a3a51)
all at ./reducedim.jl:661
unknown function (ip: 0x12b1a38d2)
default_Fun at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/constructors.jl:132
Type at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/constructors.jl:177
default_Fun at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/constructors.jl:100
Type at /Users/sheehanolver/Documents/Papers/Multivariate Orthogonal Polynomials/MultivariateTriangle/dev/ApproxFunBase/src/constructors.jl:177
unknown function (ip: 0x12b18016c)
do_call at /Users/sheehanolver/Projects/julia-1.1/src/interpreter.c:323
eval_body at /Users/sheehanolver/Projects/julia-1.1/src/interpreter.c:625
eval_body at /Users/sheehanolver/Projects/julia-1.1/src/interpreter.c:689
eval_body at /Users/sheehanolver/Projects/julia-1.1/src/interpreter.c:689
eval_body at /Users/sheehanolver/Projects/julia-1.1/src/interpreter.c:689
eval_body at /Users/sheehanolver/Projects/julia-1.1/src/interpreter.c:689
jl_interpret_toplevel_thunk_callback at /Users/sheehanolver/Projects/julia-1.1/src/interpreter.c:885
unknown function (ip: 0xfffffffffffffffe)
unknown function (ip: 0x11d5809bf)
unknown function (ip: 0xf0)
jl_interpret_toplevel_thunk at /Users/sheehanolver/Projects/julia-1.1/src/interpreter.c:894
jl_toplevel_eval_flex at /Users/sheehanolver/Projects/julia-1.1/src/toplevel.c:764
jl_toplevel_eval at /Users/sheehanolver/Projects/julia-1.1/src/toplevel.c:773 [inlined]
jl_toplevel_eval_in at /Users/sheehanolver/Projects/julia-1.1/src/toplevel.c:793
eval at ./boot.jl:328
eval_user_input at /Users/sheehanolver/Projects/julia-1.1/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:85
macro expansion at /Users/sheehanolver/Projects/julia-1.1/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:117 [inlined]
#26 at ./task.jl:259
jl_apply at /Users/sheehanolver/Projects/julia-1.1/src/./julia.h:1571 [inlined]
start_task at /Users/sheehanolver/Projects/julia-1.1/src/task.c:572
Allocations: 104721423 (Pool: 104712812; Big: 8611); GC: 116
Segmentation fault: 11
```
Any ideas @MikaelSlevinsky? I'm now using `pad(c,n,4n-3)` for the size of the array.

**Comment by MikaelSlevinsky** [comment]
Perhaps it's time to consolidate the effort to wrap the C code with https://github.com/MikaelSlevinsky/FastTransforms.jl/blob/feat-libfasttransforms/src/libfasttransforms.jl
This can be part of an optional build (for now). One thing I learned from FFTW.jl is to record the C plans with julia's garbage collector through the `finalizer` function. This may help julia know when to free memory and may be the reason for the segfaults.

**Comment by dlfivefifty** [comment]
I don’t see how a finaliser could be the problem: Julia only garbage collects objects it creates, so the plan would just be a memory leak

File: src/Disk/Disk.jl [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+*(P::TransformPlan{<:Any,<:ChebyshevDisk}, v::AbstractArray) = fromrectcfs(P.space, P.plan*v)
+*(P::ITransformPlan{<:Any,<:ChebyshevDisk}, v::AbstractArray) = P.plan*torectcfs(P.space, v)
+
+evaluate(cfs::AbstractVector, S::ChebyshevDisk, x) = evaluate(torectcfs(S,cfs), rectspace(S), ipolar(x))
+
+
+function _coefficients(disk2cxf, v̂::AbstractVector, ::ChebyshevDisk, ::ZernikeDisk)
+    F = totensor(ChebyshevDisk(), v̂)
+    F *= sqrt(convert(T,π))
```
Comment (MikaelSlevinsky) [inline src/Disk/Disk.jl]: For sin/cos Fourier series, the orthonormalization of the constant is still `1/sqrt(2pi)`.

```suggestion
    F *= sqrt(convert(T,π))
    F[:,1] *= sqrt(convert(T, 2))
```

File: src/Disk/Disk.jl [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    n = size(F,1)
+    F̌ = disk2cxf \ pad(F,n,4n-3)
+    trivec(F̌)
+end
+
+function _coefficients(disk2cxf, v::AbstractVector, ::ZernikeDisk,  ::ChebyshevDisk)
+    F̌ = tridevec(v)
+    F̌ /= sqrt(convert(T,π))
```
Comment (MikaelSlevinsky) [inline src/Disk/Disk.jl]: Same here
```suggestion
    F̌ /= sqrt(convert(T,π))
    F̌[:,1] /= sqrt(convert(T, 2))
```

File: src/c_transforms.jl [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+struct CDisk2CxfPlan
+    plan::PlanPtr
+    n::Int
+end
+
+CDisk2CxfPlan(n::Int) = CDisk2CxfPlan(c_plan_disk2cxf(n), n)
+
+function *(C::CDisk2CxfPlan, A::Matrix{Float64})
+    (size(A,1) == C.n+1 && size(A,2) == 4C.n-3) || throw(ArgumentError(A))
```
Comment (MikaelSlevinsky) [inline src/c_transforms.jl]: This still isn't right: probably
```suggestion
    (size(A,1) == C.n && size(A,2) == 4C.n-3) || throw(ArgumentError(A))
```
as below

**Comment by dlfivefifty** [comment]
Hmm you might be write its about the plan. The following seems to also trigger the bug:
```julia
julia> for _=1:1000
           CDisk2CxfPlan(1)
       end

signal (11): Segmentation fault: 11
in expression starting at no file:0
_ZN4llvm20AAResultsWrapperPass13runOnFunctionERNS_8FunctionE at /Users/solver/Projects/julia-1.1/usr/lib/libLLVM.dylib (unknown line)
_ZN4llvm13FPPassManager13runOnFunctionERNS_8FunctionE at /Users/solver/Projects/julia-1.1/usr/lib/libLLVM.dylib (unknown line)
_ZN4llvm13FPPassManager11runOnModuleERNS_6ModuleE at /Users/solver/Projects/julia-1.1/usr/lib/libLLVM.dylib (unknown line)
_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE at /Users/solver/Projects/julia-1.1/usr/lib/libLLVM.dylib (unknown line)
```

**Pull request merged**